### ISSUE TITLE
"ValueError: Autograd4bitQuantLinear() does not have a parameter or a buffer named weight." / "TypeError: LlamaForCausalLM.init() got an unexpected keyword argument 'load_in_4bit'"

### ISSUE THREAD

**Issue by enn-nafnlaus** [axolotl-ai-cloud/axolotl#259]
Goal:

Train a 3B model
... on my Linux box on GPU #0 only (CUDA_VISIBLE_DEVICES=0),
... using a dataset I built, in alpaca json format
... ideally using as little memory in training as possible (not critical here for this tiny 3B model, but important for bigger future models, for which this is a test case)
... ideally as fast as possible,
... and ideally yielding a 4-bit quantized GPTQ model

So I'm trying to follow the example and the "guide" on the axolotl project page, but there's tons of parameters and config options, and everything seems to lead down a "that's broken, with an obscure error message" road. :(  The current error:

Traceback (most recent call last):
File "/home/username/axolotl/src/axolotl/utils/models.py", line 194, in load_model
model, _ = load_llama_model_4bit_low_ram(
File "/home/username/.local/lib/python3.10/site-packages/alpaca_lora_4bit/autograd_4bit.py", line 249, in load_llama_model_4bit_low_ram
model = accelerate.load_checkpoint_and_dispatch(
File "/home/username/.local/lib/python3.10/site-packages/accelerate/big_modeling.py", line 486, in load_checkpoint_and_dispatch
load_checkpoint_in_model(
File "/home/username/.local/lib/python3.10/site-packages/accelerate/utils/modeling.py", line 1116, in load_checkpoint_in_model
set_module_tensor_to_device(model, param_name, param_device, value=param, dtype=dtype)
File "/home/username/.local/lib/python3.10/site-packages/accelerate/utils/modeling.py", line 149, in set_module_tensor_to_device
raise ValueError(f"{module} does not have a parameter or a buffer named {tensor_name}.")
ValueError: Autograd4bitQuantLinear() does not have a parameter or a buffer named weight.
Traceback (most recent call last):
File "/home/username/axolotl/src/axolotl/utils/models.py", line 194, in load_model
model, _ = load_llama_model_4bit_low_ram(
File "/home/username/.local/lib/python3.10/site-packages/alpaca_lora_4bit/autograd_4bit.py", line 249, in load_llama_model_4bit_low_ram
model = accelerate.load_checkpoint_and_dispatch(
File "/home/username/.local/lib/python3.10/site-packages/accelerate/big_modeling.py", line 486, in load_checkpoint_and_dispatch
load_checkpoint_in_model(
File "/home/username/.local/lib/python3.10/site-packages/accelerate/utils/modeling.py", line 1116, in load_checkpoint_in_model
set_module_tensor_to_device(model, param_name, param_device, value=param, dtype=dtype)
File "/home/username/.local/lib/python3.10/site-packages/accelerate/utils/modeling.py", line 149, in set_module_tensor_to_device
raise ValueError(f"{module} does not have a parameter or a buffer named {tensor_name}.")
ValueError: Autograd4bitQuantLinear() does not have a parameter or a buffer named weight.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
File "/home/username/axolotl/scripts/finetune.py", line 352, in
fire.Fire(train)
File "/home/username/.local/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
component_trace = _Fire(component, args, parsed_flag_args, context, name)
File "/home/username/.local/lib/python3.10/site-packages/fire/core.py", line 475, in _Fire
component, remaining_args = _CallAndUpdateTrace(
File "/home/username/.local/lib/python3.10/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
component = fn(*varargs, **kwargs)
File "/home/username/axolotl/scripts/finetune.py", line 251, in train
model, peft_config = load_model(
File "/home/username/axolotl/src/axolotl/utils/models.py", line 290, in load_model
model = AutoModelForCausalLM.from_pretrained(
File "/home/username/.local/lib/python3.10/site-packages/transformers/models/auto/auto_factory.py", line 467, in from_pretrained
return model_class.from_pretrained(
File "/home/username/.local/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2611, in from_pretrained
model = cls(config, *model_args, **model_kwargs)
TypeError: LlamaForCausalLM.init() got an unexpected keyword argument 'load_in_4bit'
Traceback (most recent call last):
File "/home/username/.local/bin/accelerate", line 8, in
sys.exit(main())
File "/home/username/.local/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 45, in main
args.func(args)
File "/home/username/.local/lib/python3.10/site-packages/accelerate/commands/launch.py", line 941, in launch_command
simple_launcher(args)
File "/home/username/.local/lib/python3.10/site-packages/accelerate/commands/launch.py", line 603, in simple_launcher
raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'scripts/finetune.py', 'summarize.yaml']' returned non-zero exit status 1.

For my accelerate config, while trying to find something, anything that would work, I've told it no DeepSpeed, no torch dynamo, not CPU only (of course), and have tried (for precision), "no", "fp8", and "fp16".  My yaml is attached, although as mentioned I've tried many different variants of it.

What _should_ I be doing to achieve my goals here and not get caught up in these errors?

[summarize.yaml.txt](https://github.com/OpenAccess-AI-Collective/axolotl/files/11941759/summarize.yaml.txt)

**Comment by enn-nafnlaus** [comment]
I suspect this has something to do with the peft / gptq install choice.  TL/DR:

The above errors are with having done the repository's instructions of "pip3 install -e .[gptq]".  But this command installs peft, which the instructions say is incompatible..

If I uninstall peft, then I get peft-missing errors.

One thread instead says "_pip uninstall peft and pip install .[int4]_" (int4 instead of gptq).  If I do that, it doesn't install peft, but errors switch instead to:

ModuleNotFoundError: No module named 'axolotl'

So that's no help.

Is it impossible to use gptq?

**Comment by NanoCode012** [comment]
Hello! You are correct in `pip3 install -e .[gptq]`. Where do you see that the instruct says is incompatible?

It should be possible to use gptq. However, I have not used it in a while, so some packages may have broke.

**Comment by NanoCode012** [comment]
> The above errors are with having done the repository's instructions of "pip3 install -e .[gptq]". But this command installs peft, which the instructions say is incompatible..

Hello @enn-nafnlaus , I found what you're talking about now. The reason why it said "not for gptq" is because `[gptq]` would install a specific commit of peft, so you do not need to install from source.

**Comment by enn-nafnlaus** [comment]
> Hello! You are correct in `pip3 install -e .[gptq]`. Where do you see that the instruct says is incompatible?
> 
> It should be possible to use gptq. However, I have not used it in a while, so some packages may have broke.

From my experience above, it's broken.  :(  So no possibilities for 4bit quantization then?

Honestly at this point I'd like to be able to get _anything_ working.  :(  I've got a vacation coming up in a couple days and I was hoping my model could be training while I'm gone, but it's looking decreasingly likely.

The goal of the model, beyond gaining practice in training them, is a dedicated lightweight fast text-summarizer (provided input in the form of "<#Tokens>:<Contents...>" to yield output of "<SummaryOfAround<#Tokens>InLength>"), so that text of any length could be summarized in blocks (recursively) down to something that will fit into the 2048-token contexts of LLaMA models. Thus allowing the model access to the key points of any history of any length (even entire books).

**Comment by enn-nafnlaus** [comment]
> Hello @enn-nafnlaus , sorry for late reply. I forgot to mention which lines. It's this mainly this line here as axolotl has been written for newer transformers version so this hack is needed unfortunately.
> 
> [axolotl/src/axolotl/utils/models.py](https://github.com/OpenAccess-AI-Collective/axolotl/blob/b9b7d4ce9292739d7bd3b6113e54786f45db7462/src/axolotl/utils/models.py#L213)
> 
> Line 213 in [b9b7d4c](https://github.com/OpenAccess-AI-Collective/axolotl/commit/b9b7d4ce9292739d7bd3b6113e54786f45db7462)
> 
>  load_in_4bit=cfg.load_in_4bit and cfg.adapter is not None, 
> Do note: commenting this out would not allow QLORA. Make sure that when you do pip install to use -e as stated in docs.
> 
> Edit: Feel free to discuss this in other thread which seems to be more appropriate than this.

So, this makes Lora training (which isn't my goal) using the stock model work again (it worked out of the box, so not a big gain regardless).  Training models is still broken.  From the stock example:

launch scripts/finetune.py examples/openllama-3b/config.yml

I get:

_INFO:root:Compiling torch model
INFO:root:Starting trainer...
  0%|                                                                                                                                                                                                            | 0/160428 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/home/username/axolotl/scripts/finetune.py", line 352, in <module>
    fire.Fire(train)
  File "/home/username/.local/lib/python3.10/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/username/.local/lib/python3.10/site-packages/fire/core.py", line 475, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/username/.local/lib/python3.10/site-packages/fire/core.py", line 691, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/username/axolotl/scripts/finetune.py", line 337, in train
    trainer.train(resume_from_checkpoint=resume_from_checkpoint)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/trainer.py", line 1537, in train
    return inner_training_loop(
  File "/home/username/.local/lib/python3.10/site-packages/transformers/trainer.py", line 1802, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/trainer.py", line 2647, in training_step
    loss = self.compute_loss(model, inputs)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/trainer.py", line 2672, in compute_loss
    outputs = model(**inputs)
  File "/home/username/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 690, in forward
    outputs = self.model(
  File "/home/username/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 570, in forward
    layer_outputs = torch.utils.checkpoint.checkpoint(
  File "/home/username/.local/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 249, in checkpoint
    return CheckpointFunction.apply(function, preserve, *args)
  File "/home/username/.local/lib/python3.10/site-packages/torch/autograd/function.py", line 506, in apply
    return super().apply(*args, **kwargs)  # type: ignore[misc]
  File "/home/username/.local/lib/python3.10/site-packages/torch/utils/checkpoint.py", line 107, in forward
    outputs = run_function(*args)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 566, in custom_forward
    return module(*inputs, output_attentions, None)
  File "/home/username/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/transformers/models/llama/modeling_llama.py", line 293, in forward
    hidden_states, self_attn_weights, present_key_value = self.self_attn(
  File "/home/username/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/username/.local/lib/python3.10/site-packages/accelerate/hooks.py", line 165, in new_forward
    output = old_forward(*args, **kwargs)
  File "/home/username/axolotl/src/axolotl/monkeypatch/llama_attn_hijack_xformers.py", line 91, in xformers_forward
    attn_output = xformers.ops.memory_efficient_attention(
  File "/home/username/.local/lib/python3.10/site-packages/xformers/ops/fmha/__init__.py", line 192, in memory_efficient_attention
    return _memory_efficient_attention(
  File "/home/username/.local/lib/python3.10/site-packages/xformers/ops/fmha/__init__.py", line 290, in _memory_efficient_attention
    return _memory_efficient_attention_forward(
  File "/home/username/.local/lib/python3.10/site-packages/xformers/ops/fmha/__init__.py", line 306, in _memory_efficient_attention_forward
    op = _dispatch_fw(inp)
  File "/home/username/.local/lib/python3.10/site-packages/xformers/ops/fmha/dispatch.py", line 94, in _dispatch_fw
    return _run_priority_list(
  File "/home/username/.local/lib/python3.10/site-packages/xformers/ops/fmha/dispatch.py", line 69, in _run_priority_list
    raise NotImplementedError(msg)
NotImplementedError: No operator found for `memory_efficient_attention_forward` with inputs:
     query       : shape=(1, 240, 32, 100) (torch.float16)
     key         : shape=(1, 240, 32, 100) (torch.float16)
     value       : shape=(1, 240, 32, 100) (torch.float16)
     attn_bias   : <class 'xformers.ops.fmha.attn_bias.LowerTriangularMask'>
     p           : 0.0
`cutlassF` is not supported because:
    query.shape[-1] % 8 != 0
    value.shape[-1] % 8 != 0
`flshattF` is not supported because:
    query.shape[-1] % 8 != 0
`tritonflashattF` is not supported because:
    query.shape[-1] % 8 != 0
    key.shape[-1] % 8 != 0
    value.shape[-1] % 8 != 0
    requires A100 GPU
`smallkF` is not supported because:
    dtype=torch.float16 (supported: {torch.float32})
    max(query.shape[-1] != value.shape[-1]) > 32
    attn_bias type is <class 'xformers.ops.fmha.attn_bias.LowerTriangularMask'>
    unsupported embed per head: 100
  0%|                                                                                                                                                                                                            | 0/160428 [00:01<?, ?it/s]
Traceback (most recent call last):
  File "/home/username/.local/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/home/username/.local/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 45, in main
    args.func(args)
  File "/home/username/.local/lib/python3.10/site-packages/accelerate/commands/launch.py", line 941, in launch_command
    simple_launcher(args)
  File "/home/username/.local/lib/python3.10/site-packages/accelerate/commands/launch.py", line 603, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3', 'scripts/finetune.py', 'examples/openllama-3b/config.yml']' returned non-zero exit status 1._

Note that at this point I've shut off everything that's potentially "good" just to try to get anything to work at all:

 * No Deepspeed
 * No torch compilation
 * Floating point modes (fp16, fp8) set to "No"
 * Not using the GPTQ pip3 install

And it still doesn't work.

Is there any way we can get an exact series of steps that will actually work to train a model?  Ideally one to train a 4bit GPTQ model in a fast and memory efficient manner, but at the very least, _any model_ (not Lora), at _any bit scale_, at _any training rate_?  It'd at least provide a starting point.  :)

**Comment by NanoCode012** [comment]
Hello, sorry for not checking back on this. The issue you got is due to `xformer: true`. I have added this to the Readme. Were you able to train so far?

**Comment by enn-nafnlaus** [comment]
Yep, thanks to the help on the Discord!  Removing the cache file generated by _accelerate config_, and making sure to **_not_** run _accelerate config_, was the other key.  :)  I also (very slowly) installed flash-attn===2.0.4 and set flash_attention=true in the yaml.  Then I later - once training turned out to be generating NaNs after a few hours - I had to switch the optimizer to lion_8bit.  Then when the loss rate was rising, not falling, I had to reduce the learning_rate by an order of magnitude, to 0.000001 (perhaps I could have gotten away with a less drastic cut, but that worked).  I also discovered that I was (barely) able to up micro_batch_size to 2 (barely) without running out of memory on my 24GB card training a 3GB model (note to people new to this: full-finetune VRAM requirements are staggeringly high!).

Loss has only dropped by ~50% in the first ~8500 cycles, but at least it's going down  :)

![image](https://github.com/OpenAccess-AI-Collective/axolotl/assets/116288799/b5e59836-8078-49e2-bdf1-4927ac92c427)

So the output isn't exactly useful yet, but at least I can leave it running weeks on end whenever I don't need my GPU  :)  Still need to figure out how to quantize the model to 4-bit, however. I looked into AutoGPTQ, but it's a library, not a program, so I'd have to write a program.

Anyway, it's a work in progress, and not at a showstopper at the moment  :)
